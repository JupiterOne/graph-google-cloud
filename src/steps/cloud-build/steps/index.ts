import fetch_cloud_build_triggers from './fetch-cloud-build-triggers';
import fetch_cloud_builds from './fetch-cloud-builds';

export default [fetch_cloud_build_triggers, fetch_cloud_builds];
