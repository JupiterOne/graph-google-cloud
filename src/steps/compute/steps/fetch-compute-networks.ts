import {
  GoogleCloudIntegrationStep,
  IntegrationStepContext,
} from '../../../types';
import { setNetworkPeerings, setPeeredNetworks } from '../../../utils/jobState';
import { ComputeClient } from '../client';
import {
  IngestionSources,
  STEP_COMPUTE_NETWORKS,
  ENTITY_TYPE_COMPUTE_NETWORK,
  ENTITY_CLASS_COMPUTE_NETWORK,
  ComputePermissions,
} from '../constants';
import { createComputeNetworkEntity } from '../converters';

export async function fetchComputeNetworks(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState, logger } = context;
  const client = new ComputeClient(
    {
      config: context.instance.config,
    },
    logger,
  );

  const peeredNetworks: string[] = [];

  await client.iterateNetworks(async (network) => {
    await jobState.addEntity(
      createComputeNetworkEntity(network, client.projectId),
    );

    if (network.peerings?.length) {
      await setNetworkPeerings(
        jobState,
        network.selfLink as string,
        network.peerings || [],
      );
      peeredNetworks.push(network.selfLink as string);
    }
  });

  await setPeeredNetworks(jobState, peeredNetworks);
}
export const fetchComputeNetworksStepMap: GoogleCloudIntegrationStep = {
  id: STEP_COMPUTE_NETWORKS,
  ingestionSourceId: IngestionSources.COMPUTE_NETWORKS,
  name: 'Compute Networks',
  entities: [
    {
      resourceName: 'Compute Networks',
      _type: ENTITY_TYPE_COMPUTE_NETWORK,
      _class: ENTITY_CLASS_COMPUTE_NETWORK,
    },
  ],
  relationships: [],
  executionHandler: fetchComputeNetworks,
  permissions: ComputePermissions.STEP_COMPUTE_NETWORKS,
  apis: ['compute.googleapis.com'],
};
