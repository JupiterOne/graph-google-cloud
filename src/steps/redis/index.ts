import {
  createDirectRelationship,
  getRawData,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import {
  GoogleCloudIntegrationStep,
  IntegrationStepContext,
} from '../../types';
import { STEP_COMPUTE_NETWORKS } from '../compute';
import { RedisClient } from './client';
import {
  ENTITY_CLASS_REDIS_INSTANCE,
  ENTITY_TYPE_REDIS_INSTANCE,
  STEP_REDIS_INSTANCES,
  RELATIONSHIP_TYPE_REDIS_INSTANCE_USES_NETWORK,
  STEP_CREATE_REDIS_INSTANCE_NETWORK_RELATIONSHIPS,
} from './constants';
import { ENTITY_TYPE_COMPUTE_NETWORK } from '../compute/constants';
import { createRedisInstanceEntity } from './converter';
import { redis_v1 } from 'googleapis';

export async function fetchRedisInstances(
  context: IntegrationStepContext,
): Promise<void> {
  const {
    jobState,
    instance: { config },
    logger,
  } = context;

  const client = new RedisClient({ config }, logger);

  await client.iterateRedisInstances(async (instance) => {
    const redisEntity = createRedisInstanceEntity(instance, client.projectId);
    await jobState.addEntity(redisEntity);
  });
}

export async function buildRedisInstanceUsesNetworkRelationships(
  context: IntegrationStepContext,
): Promise<void> {
  const { jobState, logger } = context;

  await jobState.iterateEntities(
    { _type: ENTITY_TYPE_REDIS_INSTANCE },
    async (redisInstanceEntity) => {
      const instance =
        getRawData<redis_v1.Schema$Instance>(redisInstanceEntity);

      if (!instance) {
        logger.warn(
          {
            _key: redisInstanceEntity._key,
          },
          'Could not find raw data on redis instance entity',
        );
        return;
      }

      const authorizedNetwork = instance.authorizedNetwork;
      if (!authorizedNetwork) {
        return;
      }

      const authorizedNetworkEntity = await jobState.findEntity(
        `https://www.googleapis.com/compute/v1/${authorizedNetwork}`,
      );
      if (!authorizedNetworkEntity) {
        return;
      }

      await jobState.addRelationship(
        createDirectRelationship({
          _class: RelationshipClass.USES,
          from: redisInstanceEntity,
          to: authorizedNetworkEntity,
        }),
      );
    },
  );
}

export const redisSteps: GoogleCloudIntegrationStep[] = [
  {
    id: STEP_REDIS_INSTANCES,
    name: 'Redis Instances',
    entities: [
      {
        resourceName: 'Redis Instance',
        _type: ENTITY_TYPE_REDIS_INSTANCE,
        _class: ENTITY_CLASS_REDIS_INSTANCE,
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchRedisInstances,
    permissions: ['redis.instances.list'],
    apis: ['redis.googleapis.com'],
  },
  {
    id: STEP_CREATE_REDIS_INSTANCE_NETWORK_RELATIONSHIPS,
    name: 'Build Redis Instance Network Relationships',
    entities: [],
    relationships: [
      {
        _class: RelationshipClass.USES,
        _type: RELATIONSHIP_TYPE_REDIS_INSTANCE_USES_NETWORK,
        sourceType: ENTITY_TYPE_REDIS_INSTANCE,
        targetType: ENTITY_TYPE_COMPUTE_NETWORK,
      },
    ],
    dependsOn: [STEP_REDIS_INSTANCES, STEP_COMPUTE_NETWORKS],
    executionHandler: buildRedisInstanceUsesNetworkRelationships,
  },
];
