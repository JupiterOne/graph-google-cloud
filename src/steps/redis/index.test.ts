import {
  StepTestConfig,
  createMockStepExecutionContext,
  executeStepWithDependencies,
} from '@jupiterone/integration-sdk-testing';
import {
  buildRedisInstanceUsesNetworkRelationships,
  fetchRedisInstances,
} from '.';
import { integrationConfig } from '../../../test/config';
import {
  Recording,
  setupGoogleCloudRecording,
  getMatchRequestsBy,
} from '../../../test/recording';
import { IntegrationConfig } from '../../types';
import { fetchComputeNetworks } from '../compute/steps/fetch-compute-networks';
import {
  ENTITY_TYPE_REDIS_INSTANCE,
  RELATIONSHIP_TYPE_REDIS_INSTANCE_USES_NETWORK,
  STEP_MEMORYSTORE_REDIS,
  STEP_MEMORYSTORE_REDIS_LOCATION,
} from './constants';
import { invocationConfig } from '../..';

const tempNewAccountConfig = {
  ...integrationConfig,
  serviceAccountKeyFile: integrationConfig.serviceAccountKeyFile.replace(
    'j1-gc-integration-dev-v2',
    'j1-gc-integration-dev-v3',
  ),
  serviceAccountKeyConfig: {
    ...integrationConfig.serviceAccountKeyConfig,
    project_id: 'j1-gc-integration-dev-v3',
  },
};

describe('#fetchRedisInstances', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchRedisInstances',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: tempNewAccountConfig,
    });

    await fetchRedisInstances(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_REDIS_INSTANCE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Database', 'DataStore', 'Host'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_redis_instance' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          redisVersion: { type: 'string' },
          hostname: { type: 'string' },
          port: { type: 'number' },
          reservedIpRange: { type: 'string' },
          state: { type: 'string' },
          tier: { type: 'string' },
          memorySizeGb: { type: 'number' },
          connectMode: { type: 'string' },
          transitEncryptionModeEnabled: { type: 'boolean' },
          statusMessage: { type: 'string' },
          encrypted: { const: null },
          classification: { const: null },
          authEnabled: { type: 'boolean' },
          alternativeLocationId: { type: 'string' },
          currentLocationId: { type: 'string' },
          locationId: { type: 'string' },
          createdOn: { type: 'number' },
          webLink: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_REDIS_INSTANCE_USES_NETWORK,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'USES' },
          _type: { const: 'google_redis_instance_uses_compute_network' },
        },
      },
    });
  });
});

describe('#buildRedisInstanceUsesNetworkRelationships', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'buildRedisInstanceUsesNetworkRelationships',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: tempNewAccountConfig,
    });

    await fetchComputeNetworks(context);
    await fetchRedisInstances(context);
    await buildRedisInstanceUsesNetworkRelationships(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_REDIS_INSTANCE_USES_NETWORK,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'USES' },
          _type: { const: 'google_redis_instance_uses_compute_network' },
        },
      },
    });
  });
});

describe(`memoryStoreRedis#${STEP_MEMORYSTORE_REDIS_LOCATION}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_MEMORYSTORE_REDIS_LOCATION, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_MEMORYSTORE_REDIS_LOCATION,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_MEMORYSTORE_REDIS_LOCATION,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});

describe(`memoryStoreRedis#${STEP_MEMORYSTORE_REDIS}`, () => {
  let recording: Recording;
  afterEach(async () => {
    if (recording) await recording.stop();
  });

  jest.setTimeout(45000);

  test(STEP_MEMORYSTORE_REDIS, async () => {
    recording = setupGoogleCloudRecording({
      name: STEP_MEMORYSTORE_REDIS,
      directory: __dirname,
      options: {
        matchRequestsBy: getMatchRequestsBy(integrationConfig),
      },
    });

    const stepTestConfig: StepTestConfig = {
      stepId: STEP_MEMORYSTORE_REDIS,
      instanceConfig: integrationConfig,
      invocationConfig: invocationConfig as any,
    };

    const result = await executeStepWithDependencies(stepTestConfig);
    expect(result).toMatchStepMetadata(stepTestConfig);
  });
});
