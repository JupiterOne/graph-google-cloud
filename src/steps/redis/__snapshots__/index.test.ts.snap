// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildRedisInstanceUsesNetworkRelationships should collect data 1`] = `
{
  "collectedEntities": [
    {
      "CIDR": null,
      "IPv4Range": undefined,
      "_class": [
        "Network",
      ],
      "_key": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "autoCreateSubnetworks": true,
            "creationTimestamp": "2021-05-31T09:21:40.121-07:00",
            "description": "Default network for the project",
            "id": "3355945658528347131",
            "kind": "compute#network",
            "name": "default",
            "networkFirewallPolicyEnforcementOrder": "AFTER_CLASSIC_FIREWALL",
            "peerings": [
              {
                "autoCreateRoutes": true,
                "exchangeSubnetRoutes": true,
                "exportCustomRoutes": false,
                "exportSubnetRoutesWithPublicIp": false,
                "importCustomRoutes": false,
                "importSubnetRoutesWithPublicIp": false,
                "name": "servicenetworking-googleapis-com",
                "network": "https://www.googleapis.com/compute/v1/projects/vc085cf1742bb325ap-tp/global/networks/servicenetworking",
                "state": "ACTIVE",
                "stateDetails": "[2022-03-15T10:41:05.027-07:00]: Connected.",
              },
            ],
            "routingConfig": {
              "routingMode": "REGIONAL",
            },
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
            "selfLinkWithId": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/3355945658528347131",
            "subnetworks": [
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-north1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west4/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west8/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east4/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/southamerica-west1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west9/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-southeast2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west4/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-central2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west6/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-south2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/southamerica-east1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-southeast1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-south1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east7/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west3/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-east1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/australia-southeast2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/australia-southeast1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west3/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/northamerica-northeast1/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/northamerica-northeast2/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast3/subnetworks/default",
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-east2/subnetworks/default",
            ],
          },
        },
      ],
      "_type": "google_compute_network",
      "autoCreateSubnetworks": true,
      "createdOn": 1622478100121,
      "description": "Default network for the project",
      "displayName": "default",
      "gatewayIPv4": undefined,
      "id": "3355945658528347131",
      "internal": true,
      "name": "default",
      "public": false,
      "routingMode": "REGIONAL",
      "subnetworks": [
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-north1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west4/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west8/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east4/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/southamerica-west1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west9/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-southeast2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west4/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-central2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west6/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-south2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/southamerica-east1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-southeast1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-south1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-east7/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west3/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-east1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/australia-southeast2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/europe-west1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/australia-southeast1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-west3/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/northamerica-northeast1/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/northamerica-northeast2/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-northeast3/subnetworks/default",
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/asia-east2/subnetworks/default",
      ],
      "webLink": "https://console.cloud.google.com/networking/networks/details/default?project=j1-gc-integration-dev-v3",
    },
    {
      "CIDR": null,
      "IPv4Range": undefined,
      "_class": [
        "Network",
      ],
      "_key": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/public-compute-app-vpc",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "autoCreateSubnetworks": false,
            "creationTimestamp": "2021-05-31T09:25:43.742-07:00",
            "id": "7739872907028377832",
            "kind": "compute#network",
            "name": "public-compute-app-vpc",
            "networkFirewallPolicyEnforcementOrder": "AFTER_CLASSIC_FIREWALL",
            "peerings": [
              {
                "autoCreateRoutes": true,
                "exchangeSubnetRoutes": true,
                "exportCustomRoutes": false,
                "exportSubnetRoutesWithPublicIp": true,
                "importCustomRoutes": false,
                "importSubnetRoutesWithPublicIp": false,
                "name": "same-project-vpc-peering-activator",
                "network": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/test-vpc-network",
                "state": "INACTIVE",
                "stateDetails": "[2021-07-02T11:37:04.253-07:00]: Peer network tried to connect and failed.",
              },
            ],
            "routingConfig": {
              "routingMode": "GLOBAL",
            },
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/public-compute-app-vpc",
            "selfLinkWithId": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/7739872907028377832",
            "subnetworks": [
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/public-compute-app-public-subnet-1",
            ],
          },
        },
      ],
      "_type": "google_compute_network",
      "autoCreateSubnetworks": false,
      "createdOn": 1622478343742,
      "description": undefined,
      "displayName": "public-compute-app-vpc",
      "gatewayIPv4": undefined,
      "id": "7739872907028377832",
      "internal": true,
      "name": "public-compute-app-vpc",
      "public": false,
      "routingMode": "GLOBAL",
      "subnetworks": [
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/public-compute-app-public-subnet-1",
      ],
      "webLink": "https://console.cloud.google.com/networking/networks/details/public-compute-app-vpc?project=j1-gc-integration-dev-v3",
    },
    {
      "CIDR": null,
      "IPv4Range": undefined,
      "_class": [
        "Network",
      ],
      "_key": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/rbs-net",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "autoCreateSubnetworks": false,
            "creationTimestamp": "2021-08-06T10:57:35.628-07:00",
            "id": "850726367629723360",
            "kind": "compute#network",
            "name": "rbs-net",
            "networkFirewallPolicyEnforcementOrder": "AFTER_CLASSIC_FIREWALL",
            "routingConfig": {
              "routingMode": "REGIONAL",
            },
            "selfLink": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/rbs-net",
            "selfLinkWithId": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/850726367629723360",
            "subnetworks": [
              "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/rbs-net-default",
            ],
          },
        },
      ],
      "_type": "google_compute_network",
      "autoCreateSubnetworks": false,
      "createdOn": 1628272655628,
      "description": undefined,
      "displayName": "rbs-net",
      "gatewayIPv4": undefined,
      "id": "850726367629723360",
      "internal": true,
      "name": "rbs-net",
      "public": false,
      "routingMode": "REGIONAL",
      "subnetworks": [
        "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/regions/us-central1/subnetworks/rbs-net-default",
      ],
      "webLink": "https://console.cloud.google.com/networking/networks/details/rbs-net?project=j1-gc-integration-dev-v3",
    },
    {
      "_class": [
        "Database",
        "DataStore",
        "Host",
      ],
      "_key": "redis:projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "authorizedNetwork": "projects/j1-gc-integration-dev-v3/global/networks/default",
            "connectMode": "PRIVATE_SERVICE_ACCESS",
            "createTime": "2022-03-15T17:40:25.188616283Z",
            "currentLocationId": "us-central1-c",
            "host": "10.0.0.131",
            "locationId": "us-central1-c",
            "memorySizeGb": 16,
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
            "nodes": [
              {
                "id": "node-0",
                "zone": "us-central1-c",
              },
            ],
            "persistenceConfig": {
              "persistenceMode": "DISABLED",
            },
            "persistenceIamIdentity": "serviceAccount:645363647218-compute@developer.gserviceaccount.com",
            "port": 6379,
            "readReplicasMode": "READ_REPLICAS_DISABLED",
            "redisVersion": "REDIS_6_X",
            "reservedIpRange": "10.0.0.128/29",
            "state": "READY",
            "tier": "BASIC",
            "transitEncryptionMode": "DISABLED",
          },
        },
      ],
      "_type": "google_redis_instance",
      "alternativeLocationId": undefined,
      "authEnabled": false,
      "classification": null,
      "connectMode": "PRIVATE_SERVICE_ACCESS",
      "createdOn": 1647366025188,
      "currentLocationId": "us-central1-c",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "encrypted": null,
      "hostname": "10.0.0.131",
      "locationId": "us-central1-c",
      "memorySizeGb": 16,
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "port": 6379,
      "redisVersion": "REDIS_6_X",
      "reservedIpRange": "10.0.0.128/29",
      "state": "READY",
      "statusMessage": undefined,
      "tier": "BASIC",
      "transitEncryptionModeEnabled": false,
      "webLink": "https://console.cloud.google.com/memorystore/redis/locations/us-central1/instances/example-redis-1/details?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "USES",
      "_fromEntityKey": "redis:projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "_key": "redis:projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1|uses|https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
      "_toEntityKey": "https://www.googleapis.com/compute/v1/projects/j1-gc-integration-dev-v3/global/networks/default",
      "_type": "google_redis_instance_uses_compute_network",
      "displayName": "USES",
    },
  ],
  "encounteredTypes": [
    "google_compute_network",
    "google_redis_instance",
    "google_redis_instance_uses_compute_network",
  ],
  "numCollectedEntities": 4,
  "numCollectedRelationships": 1,
}
`;

exports[`#fetchRedisInstances should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Database",
        "DataStore",
        "Host",
      ],
      "_key": "redis:projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "authorizedNetwork": "projects/j1-gc-integration-dev-v3/global/networks/default",
            "connectMode": "PRIVATE_SERVICE_ACCESS",
            "createTime": "2022-03-15T17:40:25.188616283Z",
            "currentLocationId": "us-central1-c",
            "host": "10.0.0.131",
            "locationId": "us-central1-c",
            "memorySizeGb": 16,
            "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
            "nodes": [
              {
                "id": "node-0",
                "zone": "us-central1-c",
              },
            ],
            "persistenceConfig": {
              "persistenceMode": "DISABLED",
            },
            "persistenceIamIdentity": "serviceAccount:645363647218-compute@developer.gserviceaccount.com",
            "port": 6379,
            "readReplicasMode": "READ_REPLICAS_DISABLED",
            "redisVersion": "REDIS_6_X",
            "reservedIpRange": "10.0.0.128/29",
            "state": "READY",
            "tier": "BASIC",
            "transitEncryptionMode": "DISABLED",
          },
        },
      ],
      "_type": "google_redis_instance",
      "alternativeLocationId": undefined,
      "authEnabled": false,
      "classification": null,
      "connectMode": "PRIVATE_SERVICE_ACCESS",
      "createdOn": 1647366025188,
      "currentLocationId": "us-central1-c",
      "displayName": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "encrypted": null,
      "hostname": "10.0.0.131",
      "locationId": "us-central1-c",
      "memorySizeGb": 16,
      "name": "projects/j1-gc-integration-dev-v3/locations/us-central1/instances/example-redis-1",
      "port": 6379,
      "redisVersion": "REDIS_6_X",
      "reservedIpRange": "10.0.0.128/29",
      "state": "READY",
      "statusMessage": undefined,
      "tier": "BASIC",
      "transitEncryptionModeEnabled": false,
      "webLink": "https://console.cloud.google.com/memorystore/redis/locations/us-central1/instances/example-redis-1/details?project=j1-gc-integration-dev-v3",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_redis_instance",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
